state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1238,13 +1222,73 @@ pub mod pallet {
 			if violating.is_empty() {
 				Ok(non_violating)
 			} else {
-				Err(BatchExecutionError::PriceLimitHit {
-					successful_swaps: non_violating.into_iter().map(|ctx| ctx.swap).collect(),
-					failed_swaps: violating.into_iter().map(|ctx| ctx.swap).collect(),
+				Err(BatchExecutionError::PriceViolation {
+					violating_swaps: violating.into_iter().map(|ctx| ctx.swap).collect(),
+					non_violating_swaps: non_violating.into_iter().map(|ctx| ctx.swap).collect(),
 				})
 			}
 		}
 
+		/// Attempts to find (and execute) a batch of swaps that wouldn't result in hitting the
+		/// price impact limit, starting with the given batch, and taking swaps out of the batch if
+		/// needed.
+		#[transactional]
+		fn execute_batch(
+			mut swaps_to_execute: Vec<Swap<T>>,
+		) -> Result<BatchExecutionOutcomes<T>, DispatchError> {
+			let mut failed_swaps = vec![];
+
+			loop {
+				if swaps_to_execute.is_empty() {
+					return Ok(BatchExecutionOutcomes { successful_swaps: vec![], failed_swaps });
+				}
+
+				match Self::try_execute_without_violations(swaps_to_execute.clone()) {
+					Ok(successful_swaps) =>
+						return Ok(BatchExecutionOutcomes { successful_swaps, failed_swaps }),
+					Err(BatchExecutionError::SwapLegFailed {
+						asset,
+						direction,
+						amount,
+						failed_swap_group,
+					}) => {
+						Self::deposit_event(Event::<T>::BatchSwapFailed {
+							asset,
+							direction,
+							amount,
+						});
```

Comment:
It might we worth adjusting how we communicate this: the swaps have not longer actually failed, they might be retried. Maybe we don't need this event at all any more?